src/garage/tf/algos/npo2.py

```diff
@@ -0,0 +1,562 @@
+
+    def _build_entropy_term(self, i):
+        """Build policy entropy tensor.
+
+        Args:
+            i (namedtuple): Collection of variables to compute policy loss.
+
+        Returns:
+            tf.Tensor: Policy entropy.
+
+        """
+        pol_dist = self.policy.distribution
+
+        with tf.name_scope('policy_entropy'):
+            if self._use_neg_logli_entropy:
+                policy_entropy = -pol_dist.log_prob(i.action_var,
+                                                    name='policy_log_likeli')
+            else:
+                policy_entropy = pol_dist.entropy()
+
+            # This prevents entropy from becoming negative for small policy std
+            if self._use_softplus_entropy:
+                policy_entropy = tf.nn.softplus(policy_entropy)
+
+            if self._stop_entropy_gradient:
+                policy_entropy = tf.stop_gradient(policy_entropy)
+
+        if not self.policy.recurrent:
+            sparse_policy_entropy = tf.RaggedTensor.from_row_lengths(
+                policy_entropy, i.length_var).to_tensor()
+            paddings = [[
+                0, 0
+            ], [0, self.max_path_length - tf.shape(sparse_policy_entropy)[1]]]
+            policy_entropy = tf.pad(sparse_policy_entropy, paddings,
+                                    'CONSTANT')
+
+        self._f_policy_entropy = compile_function(
+            flatten_inputs(self._policy_opt_inputs), policy_entropy)
+
+        return policy_entropy
+
+    def _fit_baseline_with_data(self, samples_data):
+        """Update baselines from samples.
+
+        Args:
+            samples_data (dict): Processed sample data.
+                See process_samples() for details.
+
+        """
+        policy_opt_input_values = self._policy_opt_input_values(samples_data)
+
+        # Augment reward from baselines
+        rewards_tensor = self._f_rewards(*policy_opt_input_values)
+        returns_tensor = self._f_returns(*policy_opt_input_values)
+        returns_tensor = np.squeeze(returns_tensor, -1)
+
+        paths = samples_data['paths']
+        valids = samples_data['valids']
+
+        # Recompute parts of samples_data
+        aug_rewards = []
+        aug_returns = []
+        for rew, ret, val, path in zip(rewards_tensor, returns_tensor, valids,
+                                       paths):
+            path['rewards'] = rew[val.astype(np.bool)]
+            path['returns'] = ret[val.astype(np.bool)]
+            aug_rewards.append(path['rewards'])
+            aug_returns.append(path['returns'])
+        samples_data['rewards'] = np_tensor_utils.pad_tensor_n(
+            aug_rewards, self.max_path_length)
+        samples_data['returns'] = np_tensor_utils.pad_tensor_n(
+            aug_returns, self.max_path_length)
+
+        # Fit baseline
+        logger.log('Fitting baseline...')
+        if hasattr(self.baseline, 'fit_with_samples'):
+            self.baseline.fit_with_samples(paths, samples_data)
```

Comment:
we don't have any baselines that implement `fit_with_samples`, so i think you can safely remove this check

Reply:
Ok